From b681bb30541a19ff41ad4a32ed08c2dc759c46f6 Mon Sep 17 00:00:00 2001
From: Chris Hofer <christian.hofer@codasip.com>
Date: Thu, 24 Jul 2025 12:41:54 +0200
Subject: [PATCH 1/2] syscall: Replace argument casts with uintptr_t

The syscall arguements of AIO syscalls can be pointers. As pointers
do not fit into longs, turn the casts into uintptr_t ones.

Upstream-Status: Inappropriate [CHERI-specific]
Signed-off-by: Chris Hofer <christian.hofer@codasip.com>
---
 src/io_pgetevents.c |  5 +++--
 src/syscall.h       | 15 ++++++++-------
 2 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/src/io_pgetevents.c b/src/io_pgetevents.c
index e6b0614..c3ede96 100644
--- a/src/io_pgetevents.c
+++ b/src/io_pgetevents.c
@@ -18,6 +18,7 @@
  */
 #include <libaio.h>
 #include <errno.h>
+#include <stdint.h>
 #include <stdlib.h>
 #include <time.h>
 #include <signal.h>
@@ -34,14 +35,14 @@ int io_pgetevents(io_context_t ctx, long min_nr, long nr,
 		sigset_t *sigmask)
 {
 	struct {
-		unsigned long ss;
+		uintptr_t ss;
 		unsigned long ss_len;
 	} data;
 
 	if (aio_ring_is_empty(ctx, timeout))
 		return 0;
 
-	data.ss = (unsigned long)sigmask;
+	data.ss = (uintptr_t)sigmask;
 	data.ss_len = _NSIG / 8;
 	return __io_pgetevents(ctx, min_nr, nr, events, timeout, &data);
 }
diff --git a/src/syscall.h b/src/syscall.h
index 8cf8470..ca80546 100644
--- a/src/syscall.h
+++ b/src/syscall.h
@@ -1,3 +1,4 @@
+#include <stdint.h>
 #include <sys/syscall.h>
 #include <unistd.h>
 #include <errno.h>
@@ -53,27 +54,27 @@
 
 #define io_syscall1(type,fname,sname,type1,arg1) \
 type fname(type1 arg1) \
-_body_io_syscall(sname, (long)arg1)
+_body_io_syscall(sname, (uintptr_t)arg1)
 
 #define io_syscall2(type,fname,sname,type1,arg1,type2,arg2) \
 type fname(type1 arg1,type2 arg2) \
-_body_io_syscall(sname, (long)arg1, (long)arg2)
+_body_io_syscall(sname, (uintptr_t)arg1, (uintptr_t)arg2)
 
 #define io_syscall3(type,fname,sname,type1,arg1,type2,arg2,type3,arg3) \
 type fname(type1 arg1,type2 arg2,type3 arg3) \
-_body_io_syscall(sname, (long)arg1, (long)arg2, (long)arg3)
+_body_io_syscall(sname, (uintptr_t)arg1, (uintptr_t)arg2, (uintptr_t)arg3)
 
 #define io_syscall4(type,fname,sname,type1,arg1,type2,arg2,type3,arg3,type4,arg4) \
 type fname (type1 arg1, type2 arg2, type3 arg3, type4 arg4) \
-_body_io_syscall(sname, (long)arg1, (long)arg2, (long)arg3, (long)arg4)
+_body_io_syscall(sname, (uintptr_t)arg1, (uintptr_t)arg2, (uintptr_t)arg3, (uintptr_t)arg4)
 
 #define io_syscall5(type,fname,sname,type1,arg1,type2,arg2,type3,arg3,type4,arg4, type5,arg5) \
 type fname (type1 arg1,type2 arg2,type3 arg3,type4 arg4,type5 arg5) \
-_body_io_syscall(sname, (long)arg1, (long)arg2, (long)arg3, (long)arg4, (long)arg5)
+_body_io_syscall(sname, (uintptr_t)arg1, (uintptr_t)arg2, (uintptr_t)arg3, (uintptr_t)arg4, (uintptr_t)arg5)
 
 #define io_syscall6(type,fname,sname,type1,arg1,type2,arg2,type3,arg3, \
 		type4,arg4,type5,arg5,type6,arg6) \
 type fname (type1 arg1,type2 arg2,type3 arg3,type4 arg4,type5 arg5, \
 		type6 arg6) \
-_body_io_syscall(sname, (long)arg1, (long)arg2, (long)arg3, (long)arg4, \
-               (long)arg5, (long)arg6)
+_body_io_syscall(sname, (uintptr_t)arg1, (uintptr_t)arg2, (uintptr_t)arg3, (uintptr_t)arg4, \
+               (uintptr_t)arg5, (uintptr_t)arg6)
-- 
2.47.1

